can only join a child process

<!-- Welcome and thank you for helping us make Seldon Core better!

To help us address your issue, please provide us as much of the information requested below as possble. Thanks! -->


## Describe the bug
<!-- A clear and concise description of what the bug is. -->
When deleting xgboost rest model server,  I got:
```
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Error in atexit._run_exitfuncs:
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Traceback (most recent call last):
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/util.py", line 322, in _exit_function
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     p.join()
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/process.py", line 138, in join
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     assert self._parent_pid == os.getpid(), 'can only join a child process'
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier AssertionError: can only join a child process
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Error in atexit._run_exitfuncs:
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Traceback (most recent call last):
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/util.py", line 322, in _exit_function
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     p.join()
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/process.py", line 138, in join
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     assert self._parent_pid == os.getpid(), 'can only join a child process'
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier AssertionError: can only join a child process
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Error in atexit._run_exitfuncs:
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Traceback (most recent call last):
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/util.py", line 322, in _exit_function
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     p.join()
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/process.py", line 138, in join
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     assert self._parent_pid == os.getpid(), 'can only join a child process'
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier AssertionError: can only join a child process
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Error in atexit._run_exitfuncs:
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Traceback (most recent call last):
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/util.py", line 322, in _exit_function
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     p.join()
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/process.py", line 138, in join
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     assert self._parent_pid == os.getpid(), 'can only join a child process'
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier AssertionError: can only join a child process
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Error in atexit._run_exitfuncs:
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Traceback (most recent call last):
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/util.py", line 322, in _exit_function
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     p.join()
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/process.py", line 138, in join
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     assert self._parent_pid == os.getpid(), 'can only join a child process'
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier AssertionError: can only join a child process
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Error in atexit._run_exitfuncs:
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Traceback (most recent call last):
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/util.py", line 322, in _exit_function
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     p.join()
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/process.py", line 138, in join
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     assert self._parent_pid == os.getpid(), 'can only join a child process'
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier AssertionError: can only join a child process
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Error in atexit._run_exitfuncs:
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier Traceback (most recent call last):
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/util.py", line 322, in _exit_function
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     p.join()
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier   File "/opt/conda/lib/python3.7/multiprocessing/process.py", line 138, in join
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier     assert self._parent_pid == os.getpid(), 'can only join a child process'
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier AssertionError: can only join a child process
xgboost-rest-xgboost-rest-0-classifier-c965c6dfb-qbhx9 classifier [2020-07-08 23:16:33 +0000] [1] [INFO] Shutting down: Master
```

Deployment file:
```
apiVersion: machinelearning.seldon.io/v1alpha2
kind: SeldonDeployment
metadata:
  name: xgboost-rest
  namespace: team-xxxx
  labels:
    seldon.io/controller-id: ''
spec:
  name: xgboost-rest
  predictors:
  - annotations:
      project_name: seldon-benchmark
    componentSpecs:
    - spec:
        containers:
        - name: classifier
          image: seldonio/xgboostserver_rest:0.4
          imagePullPolicy: Always
          resources:
            requests:
              memory: 1Gi
              cpu: 4
            limits:
              memory: 4Gi
              cpu: 6
          env:
          - name: SELDON_LOG_LEVEL
            value: DEBUG
          - name: GUNICORN_MAX_REQUESTS
            value: "10000"
          - name: GUNICORN_MAX_REQUESTS_JITTER
            value: "1000"
          - name: GUNICORN_WORKERS
            value: "7"
          - name: KMP_AFFINITY
            value: "granularity=fine,verbose,compact,1,0"
          - name: KMP_BLOCKTIME
            value: "1"
          - name: OMP_NUM_THREADS
            value: "8"
        terminationGracePeriodSeconds: 2
    graph:
      children: []
      implementation: XGBOOST_SERVER
      modelUri: gs://ml_models_test/v0002
      serviceAccountName: seldon-core-manager-key
      name: classifier
      endpoint:
        type: REST
    name: xgboost-rest
    replicas: 1
    svcOrchSpec:
      resources:
        requests:
          cpu: 2
          memory: 2Gi
      env:
      - name: SELDON_LOG_LEVEL
        value: DEBUG
```

## To reproduce
<!-- Steps required to reproduce the issue. For example:
1. define model ...
2. build image ...  (especially what wrapper version is used)
3. deploy ...
-->

## Expected behaviour
<!-- A clear and concise description of what you expected to happen. -->


## Environment
<!-- Description of environment -->
gke 
k8s client 1.18
k8s server 1.15
seldon-core: 1.2.1

<!-- You Can fill it manually or paste the output of the command below:

* Cloud Provider: [e.g. GKE, AWS, Bare Metal, Kind, Minikube]
* Kubernetes Cluster Version [Output of `kubectl version`] 
* Deployed Seldon System Images: [Output of `kubectl get --namespace seldon-system deploy seldon-controller-manager -o yaml  | grep seldonio`]

Alternatively run `echo "#### Kubernetes version:\n $(kubectl version) \n\n#### Seldon Images:\n$(kubectl get --namespace seldon-system deploy seldon-controller-manager -o yaml  | grep seldonio)"`
-->

## Model Details <!-- If the issue is with your deployed model you can also provide the following for fulll insights -->
* Images of your model: [Output of: `kubectl get seldondeployment -n <yourmodelnamespace> <seldondepname> -o yaml | grep image:` where `<yourmodelnamespace>`]
* Logs of your model: [You can get the logs of your model by running `kubectl logs -n <yourmodelnamespace> <seldonpodname> <container>`]

--- a/python/seldon_core/app.py
+++ b/python/seldon_core/app.py
@@ -1,10 +1,18 @@
 import os
 import logging
+import atexit
 
+from multiprocessing.util import _exit_function
 from typing import Dict, Union
 from gunicorn.app.base import BaseApplication
 
 logger = logging.getLogger(__name__)
+
+
+def post_worker_init(worker):
+    # Remove the atexit handler set up by the parent process
+    # https://github.com/benoitc/gunicorn/issues/1391#issuecomment-467010209
+    atexit.unregister(_exit_function)
 
 
 def accesslog(log_level: str) -> Union[str, None]:

--- a/python/seldon_core/microservice.py
+++ b/python/seldon_core/microservice.py
@@ -20,6 +20,7 @@ from seldon_core.app import (
     UserModelApplication,
     accesslog,
     threads,
+    post_worker_init,
 )
 
 logger = logging.getLogger(__name__)
@@ -53,18 +54,23 @@ def start_servers(
        Auxilary flask process
 
     """
-    p2 = mp.Process(target=target2)
-    p2.daemon = True
-    p2.start()
+    p2 = None
+    if target2:
+        p2 = mp.Process(target=target2, daemon=True)
+        p2.start()
 
-    p3 = mp.Process(target=metrics_target)
-    p3.daemon = True
-    p3.start()
+    p3 = None
+    if metrics_target:
+        p3 = mp.Process(target=metrics_target, daemon=True)
+        p3.start()
 
     target1()
 
-    p2.join()
-    p3.join()
+    if p2:
+        p2.join()
+
+    if p3:
+        p3.join()
 
 
 def parse_parameters(parameters: Dict) -> Dict:
@@ -391,6 +397,7 @@ def main():
                     "workers": args.workers,
                     "max_requests": args.max_requests,
                     "max_requests_jitter": args.max_requests_jitter,
+                    "post_worker_init": post_worker_init,
                 }
                 app = seldon_microservice.get_rest_microservice(
                     user_object, seldon_metrics
@@ -452,6 +459,7 @@ def main():
                 "timeout": 5000,
                 "max_requests": args.max_requests,
                 "max_requests_jitter": args.max_requests_jitter,
+                "post_worker_init": post_worker_init,
             }
             StandaloneApplication(app, options=options).run()
 
@@ -465,7 +473,7 @@ def main():
     else:
         server2_func = None
 
-        logger.info("Starting servers")
+    logger.info("Starting servers")
     start_servers(server1_func, server2_func, metrics_server_func)
 
 

